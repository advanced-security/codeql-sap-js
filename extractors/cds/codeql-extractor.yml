name: "cds"
display_name: "SAP CAP CDS"
version: 0.0.1
column_kind: "utf16"
file_types:
  - name: cds
    display_name: SAP CAP CDS files
    extensions:
      - .cds
build_modes:
  - none