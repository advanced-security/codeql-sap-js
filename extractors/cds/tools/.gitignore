# Ignore files create just for debugging the CDS extractor.
debug/

# Override the repository-level .gitignore to explicitly include the "dist" and
#"node_modules" sub-directories, which are used to store the CDS extractor JS build
#files and their Node.js dependencies, respectively.
!dist/
!node_modules/

