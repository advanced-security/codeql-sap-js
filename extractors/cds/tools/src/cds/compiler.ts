import { execFileSync, spawnSync, SpawnSyncReturns } from 'child_process';
import { resolve } from 'path';

import { fileExists, dirExists, recursivelyRenameJsonFiles } from '../filesystem';

/**
 * Result of a CDS compilation
 */
export interface CdsCompilationResult {
  success: boolean;
  message?: string;
  outputPath?: string;
}

/**
 * Determine the `cds` command to use based on the environment.
 * @returns A string representing the CLI command to run to invoke the
 * CDS compiler.
 */
export function determineCdsCommand(): string {
  let cdsCommand = 'cds';
  // TODO : create a mapping of project sub-directories to the correct
  // cds command to use, which will also determine the version of the cds
  // compiler that will be used for compiling `.cds` files to `.cds.json`
  // files for that sub-directory / project.
  try {
    execFileSync('cds', ['--version'], { stdio: 'ignore' });
  } catch {
    // If 'cds' command is not available, use npx to run it
    cdsCommand = 'npx -y --package @sap/cds-dk cds';
  }
  return cdsCommand;
}

/**
 * Compile a CDS file to JSON
 * @param cdsFilePath Path to the CDS file
 * @param sourceRoot The source root directory
 * @param cdsCommand The CDS command to use
 * @returns Result of the compilation
 */
export function compileCdsToJson(
  cdsFilePath: string,
  sourceRoot: string,
  cdsCommand: string,
): CdsCompilationResult {
  try {
    const resolvedCdsFilePath = resolve(cdsFilePath);
    if (!fileExists(resolvedCdsFilePath)) {
      throw new Error(`Expected CDS file '${resolvedCdsFilePath}' does not exist.`);
    }

    const cdsJsonOutPath = `${resolvedCdsFilePath}.json`;
    console.log(`Processing CDS file ${resolvedCdsFilePath} to ${cdsJsonOutPath} ...`);

    const result: SpawnSyncReturns<Buffer> = spawnSync(
      cdsCommand,
      [
        'compile',
        resolvedCdsFilePath,
        '--to',
        'json',
        '--dest',
        cdsJsonOutPath,
        '--locations',
        '--log-level',
        'warn',
      ],
      { cwd: sourceRoot, shell: true, stdio: 'pipe' },
    );

    if (result.error) {
      throw new Error(`Error executing CDS compiler: ${result.error.message}`);
    }

    if (result.status !== 0) {
      throw new Error(
        `Could not compile the file ${resolvedCdsFilePath}.\nReported error(s):\n\`\`\`\n${
          result.stderr?.toString() || 'Unknown error'
        }\n\`\`\``,
      );
    }

    if (!fileExists(cdsJsonOutPath) && !dirExists(cdsJsonOutPath)) {
      throw new Error(
        `CDS source file '${resolvedCdsFilePath}' was not compiled to JSON. This is likely because the file does not exist or is not a valid CDS file.`,
      );
    }

    // Handle directory output if the CDS compiler generated a directory
    if (dirExists(cdsJsonOutPath)) {
      console.log(`CDS compiler generated JSON to output directory: ${cdsJsonOutPath}`);
      // Recursively rename all .json files to have a .cds.json extension
      recursivelyRenameJsonFiles(cdsJsonOutPath);
    } else {
      console.log(`CDS compiler generated JSON to file: ${cdsJsonOutPath}`);
    }

    return { success: true, outputPath: cdsJsonOutPath };
  } catch (error) {
    return { success: false, message: String(error) };
  }
}
