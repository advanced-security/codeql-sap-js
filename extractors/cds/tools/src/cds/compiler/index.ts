export { determineCdsCommand, getCommandAnalysisForDebug, resetCdsCommandCache } from './command';
export type { CdsCommandAnalysis } from './command';
export { compileCdsToJson } from './compile';
export {
  executeCompilationTasks,
  generateStatusReport,
  orchestrateCompilation,
  planCompilationTasks,
} from './graph';
export { findProjectForCdsFile } from './project';
export type {
  AlternativeCdsCommand,
  CdsCompilationResult,
  CompilationAttempt,
  CompilationStatus,
  CompilationTask,
  EnhancedCompilationConfig,
} from './types';
export { getCdsVersion } from './version';
