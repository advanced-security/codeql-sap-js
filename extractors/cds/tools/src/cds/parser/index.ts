export { writeParserDebugInfo } from './debug';
export { buildCdsProjectDependencyGraph, buildEnhancedCdsProjectDependencyGraph } from './graph';
export * from './functions';
export * from './types';
