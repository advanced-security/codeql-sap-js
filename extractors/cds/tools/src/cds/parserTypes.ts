/** Context for CDS access control. */
export interface CdsAccessControl {
  /** Source file where the access control is defined. */
  sourceFile: string;

  /** Service or entity being restricted. */
  target: string;

  /** Type of restriction (e.g. 'requires', 'grant'). */
  type: string;

  /** Authorization value. */
  value: unknown;
}

/** Represents a CDS annotation. */
export interface CdsAnnotation {
  /** Name of the annotation. */
  name: string;

  /** Source file where the annotation is defined. */
  sourceFile: string;

  /** Value of the annotation. */
  value: unknown;
}

/** Represents a CDS entity definition. */
export interface CdsEntity {
  /** Annotations attached to the entity. */
  annotations: CdsAnnotation[];

  /** Parent entity or aspect if extended. */
  extends?: string;

  /** Full qualified name including namespace. */
  fqn: string;

  /** Name of the entity. */
  name: string;

  /** Properties of the entity. */
  properties: CdsProperty[];

  /** Source file where the entity is defined. */
  sourceFile: string;
}

/** Represents an entity exposed in a CDS service. */
export interface CdsExposedEntity {
  /** Annotations attached to the exposed entity. */
  annotations: CdsAnnotation[];

  /** Whether this is a projection. */
  isProjection: boolean;

  /** Name of the exposed entity in the service. */
  name: string;

  /** Source entity that is being exposed. */
  sourceEntity: string;
}

/** Represents an import reference in a CDS file. */
export interface CdsImport {
  /** Whether the import is from a module (node_modules). */
  isModule: boolean;

  /** Whether the import is relative. */
  isRelative: boolean;

  /** Path to the imported resource. */
  path: string;

  /** Original import statement. */
  statement: string;
}

/** Results of parsing CDS content. */
export interface CdsParseResult {
  /** List of access controls found. */
  accessControls: CdsAccessControl[];

  /** List of context blocks. */
  contexts: {
    entities: CdsEntity[];
    name: string;
    services: CdsService[];
  }[];

  /** List of entities found. */
  entities: CdsEntity[];

  /** Errors encountered during parsing. */
  errors: string[];

  /** List of imports found. */
  imports: CdsImport[];

  /** Namespace declared in the file. */
  namespace?: string;

  /** List of services found. */
  services: CdsService[];
}

/** Represents a property of a CDS entity. */
export interface CdsProperty {
  /** Annotations attached to the property. */
  annotations: CdsAnnotation[];

  /** Cardinality if this is an association. */
  cardinality?: 'one' | 'many';

  /** Whether this property is an association. */
  isAssociation: boolean;

  /** Whether this is a composition. */
  isComposition: boolean;

  /** Whether this property is a key. */
  isKey: boolean;

  /** Name of the property. */
  name: string;

  /** Target entity if this is an association. */
  target?: string;

  /** Data type of the property. */
  type: string;
}

/** Represents a simplified package.json file structure with only the fields we need */
export interface PackageJson {
  /** The name of the package */
  name?: string;

  /** The version of the package */
  version?: string;

  /** Production dependencies */
  dependencies?: Record<string, string>;

  /** Development dependencies */
  devDependencies?: Record<string, string>;

  /** All other fields in package.json */
  [key: string]: unknown;
}

/** File cache to avoid reading the same file multiple times */
export interface FileCache {
  /** Map of file paths to their content */
  fileContents: Map<string, string>;

  /** Map of file paths to their parsed package.json content */
  packageJsonCache: Map<string, PackageJson>;

  /** Map of file paths to their parsed CDS content */
  cdsParseCache: Map<string, CdsParseResult>;
}

/** Represents a CDS project with its directory and associated files. */
export interface CdsProject {
  /** All CDS files within this project. */
  cdsFiles: string[];

  /** Dependencies on other CDS projects. */
  dependencies?: CdsProject[];

  /** The package.json content if available. */
  packageJson?: PackageJson;

  /** The directory path of the project. */
  projectDir: string;
}

/** Represents a CDS service definition. */
export interface CdsService {
  /** Annotations attached to the service. */
  annotations: CdsAnnotation[];

  /** Exposed entities in this service. */
  entities: CdsExposedEntity[];

  /** Full qualified name including namespace. */
  fqn: string;

  /** Name of the service. */
  name: string;

  /** Source file where the service is defined. */
  sourceFile: string;
}
