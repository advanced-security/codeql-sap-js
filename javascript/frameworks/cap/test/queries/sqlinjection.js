import cds from '@sap/cds'
const { Books } = cds.entities ('sap.capire.bookshop')

class SampleVulnService extends cds.ApplicationService { init(){

  // contains a sample sql injection
  this.on ('submitOrder', async req => {
    const {book,quantity} = req.data

    let {stock} = await SELECT `stock` .from (Books,book) //alert?

    let query = SELECT.from `Books` .where (`ID=${book}`)
    let books = await cds.db.run (query) //alert

    let books11 = await SELECT.from `Books` .where (`ID=${book}`) //alert

    let query2 = SELECT.from `Books` .where ('ID='+book)
    let books2 = await cds.db.run (query2) //alert

    let books22 = await SELECT.from `Books` .where ('ID='+book) //alert

    let books3 = await SELECT.from `Books` .where `ID=${book}` //safe

    let id=2
    let books33 = await SELECT.from `Books` .where ('ID='+id) //safe

    let cqn = CQL`SELECT col1, col2, col3 from Books` + book
    let books222 = await cds.db.run (cqn) //alert

    let cqn1 = cds.parse.cql (`SELECT * from Books`+ book) 
    let books111 = await cds.db.run (cqn1) //alert
  })

  return super.init()
}}
export { SampleVulnService }
