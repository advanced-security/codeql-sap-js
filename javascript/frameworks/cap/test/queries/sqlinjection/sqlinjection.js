import cds from '@sap/cds'
const { Books } = cds.entities('sap.capire.bookshop')

class SampleVulnService extends cds.ApplicationService {
  init() {
    // contains a sample CQL injection
    this.on('submitOrder', async req => {
      const { book, quantity } = req.data

      let { stock } = await SELECT`stock`.from(Books, book)

      let query = SELECT.from`Books`.where(`ID=${book}`)
      let books = await cds.db.run(query) // CQL injection alert

      let books11 = await SELECT.from`Books`.where(`ID=${book}`) // CQL injection alert

      let query2 = SELECT.from`Books`.where('ID=' + book)
      let books2 = await cds.db.run(query2) // CQL injection alert

      let books22 = await SELECT.from`Books`.where('ID=' + book) // CQL injection alert

      let books3 = await SELECT.from`Books`.where`ID=${book}` //safe

      let id = 2
      let books33 = await SELECT.from`Books`.where('ID=' + id) //safe

      let cqn = CQL`SELECT col1, col2, col3 from Books` + book
      let books222 = await cds.db.run(cqn) // CQL injection alert

      let cqn1 = cds.parse.cql(`SELECT * from Books` + book)
      let books111 = await cds.db.run(cqn1) // CQL injection alert

      const pg = require("pg"),
        pool = new pg.Pool(config);
      pool.query(req.params.category, [], function (err, results) { // non-CQL injection alert from CAP source
        // process results
      });

      const app = require("express")();
      app.get("search", function handler(req2, res) {
        pool.query(req2.params.category, [], function (err, results) { // non-CQL injection alert from non-CAP source
          // process results
        });
      });

      return super.init()
    })
  }
}
export { SampleVulnService }