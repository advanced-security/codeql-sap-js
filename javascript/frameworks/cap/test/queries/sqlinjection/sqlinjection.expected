nodes
| sqlinjection.js:7:33:7:35 | req |
| sqlinjection.js:7:33:7:35 | req |
| sqlinjection.js:8:11:8:25 | {book,quantity} |
| sqlinjection.js:8:11:8:36 | book |
| sqlinjection.js:8:12:8:15 | book |
| sqlinjection.js:8:29:8:31 | req |
| sqlinjection.js:8:29:8:36 | req.data |
| sqlinjection.js:12:9:12:57 | query |
| sqlinjection.js:12:17:12:57 | SELECT. ... book}`) |
| sqlinjection.js:12:45:12:56 | `ID=${book}` |
| sqlinjection.js:12:51:12:54 | book |
| sqlinjection.js:13:35:13:39 | query |
| sqlinjection.js:13:35:13:39 | query |
| sqlinjection.js:15:25:15:65 | SELECT. ... book}`) |
| sqlinjection.js:15:25:15:65 | SELECT. ... book}`) |
| sqlinjection.js:15:53:15:64 | `ID=${book}` |
| sqlinjection.js:15:59:15:62 | book |
| sqlinjection.js:17:9:17:56 | query2 |
| sqlinjection.js:17:18:17:56 | SELECT. ... '+book) |
| sqlinjection.js:17:46:17:55 | 'ID='+book |
| sqlinjection.js:17:52:17:55 | book |
| sqlinjection.js:18:36:18:41 | query2 |
| sqlinjection.js:18:36:18:41 | query2 |
| sqlinjection.js:20:25:20:63 | SELECT. ... '+book) |
| sqlinjection.js:20:25:20:63 | SELECT. ... '+book) |
| sqlinjection.js:20:53:20:62 | 'ID='+book |
| sqlinjection.js:20:59:20:62 | book |
| sqlinjection.js:27:9:27:60 | cqn |
| sqlinjection.js:27:15:27:60 | CQL`SEL ...  + book |
| sqlinjection.js:27:57:27:60 | book |
| sqlinjection.js:28:38:28:40 | cqn |
| sqlinjection.js:28:38:28:40 | cqn |
| sqlinjection.js:30:9:30:58 | cqn1 |
| sqlinjection.js:30:16:30:58 | cds.par ... + book) |
| sqlinjection.js:30:31:30:57 | `SELECT ... `+ book |
| sqlinjection.js:30:54:30:57 | book |
| sqlinjection.js:31:38:31:41 | cqn1 |
| sqlinjection.js:31:38:31:41 | cqn1 |
edges
| sqlinjection.js:7:33:7:35 | req | sqlinjection.js:8:29:8:31 | req |
| sqlinjection.js:7:33:7:35 | req | sqlinjection.js:8:29:8:31 | req |
| sqlinjection.js:8:11:8:25 | {book,quantity} | sqlinjection.js:8:12:8:15 | book |
| sqlinjection.js:8:11:8:36 | book | sqlinjection.js:12:51:12:54 | book |
| sqlinjection.js:8:11:8:36 | book | sqlinjection.js:15:59:15:62 | book |
| sqlinjection.js:8:11:8:36 | book | sqlinjection.js:17:52:17:55 | book |
| sqlinjection.js:8:11:8:36 | book | sqlinjection.js:20:59:20:62 | book |
| sqlinjection.js:8:11:8:36 | book | sqlinjection.js:27:57:27:60 | book |
| sqlinjection.js:8:11:8:36 | book | sqlinjection.js:30:54:30:57 | book |
| sqlinjection.js:8:12:8:15 | book | sqlinjection.js:8:11:8:36 | book |
| sqlinjection.js:8:29:8:31 | req | sqlinjection.js:8:29:8:36 | req.data |
| sqlinjection.js:8:29:8:36 | req.data | sqlinjection.js:8:11:8:25 | {book,quantity} |
| sqlinjection.js:12:9:12:57 | query | sqlinjection.js:13:35:13:39 | query |
| sqlinjection.js:12:9:12:57 | query | sqlinjection.js:13:35:13:39 | query |
| sqlinjection.js:12:17:12:57 | SELECT. ... book}`) | sqlinjection.js:12:9:12:57 | query |
| sqlinjection.js:12:45:12:56 | `ID=${book}` | sqlinjection.js:12:17:12:57 | SELECT. ... book}`) |
| sqlinjection.js:12:51:12:54 | book | sqlinjection.js:12:45:12:56 | `ID=${book}` |
| sqlinjection.js:15:53:15:64 | `ID=${book}` | sqlinjection.js:15:25:15:65 | SELECT. ... book}`) |
| sqlinjection.js:15:53:15:64 | `ID=${book}` | sqlinjection.js:15:25:15:65 | SELECT. ... book}`) |
| sqlinjection.js:15:59:15:62 | book | sqlinjection.js:15:53:15:64 | `ID=${book}` |
| sqlinjection.js:17:9:17:56 | query2 | sqlinjection.js:18:36:18:41 | query2 |
| sqlinjection.js:17:9:17:56 | query2 | sqlinjection.js:18:36:18:41 | query2 |
| sqlinjection.js:17:18:17:56 | SELECT. ... '+book) | sqlinjection.js:17:9:17:56 | query2 |
| sqlinjection.js:17:46:17:55 | 'ID='+book | sqlinjection.js:17:18:17:56 | SELECT. ... '+book) |
| sqlinjection.js:17:52:17:55 | book | sqlinjection.js:17:46:17:55 | 'ID='+book |
| sqlinjection.js:20:53:20:62 | 'ID='+book | sqlinjection.js:20:25:20:63 | SELECT. ... '+book) |
| sqlinjection.js:20:53:20:62 | 'ID='+book | sqlinjection.js:20:25:20:63 | SELECT. ... '+book) |
| sqlinjection.js:20:59:20:62 | book | sqlinjection.js:20:53:20:62 | 'ID='+book |
| sqlinjection.js:27:9:27:60 | cqn | sqlinjection.js:28:38:28:40 | cqn |
| sqlinjection.js:27:9:27:60 | cqn | sqlinjection.js:28:38:28:40 | cqn |
| sqlinjection.js:27:15:27:60 | CQL`SEL ...  + book | sqlinjection.js:27:9:27:60 | cqn |
| sqlinjection.js:27:57:27:60 | book | sqlinjection.js:27:15:27:60 | CQL`SEL ...  + book |
| sqlinjection.js:30:9:30:58 | cqn1 | sqlinjection.js:31:38:31:41 | cqn1 |
| sqlinjection.js:30:9:30:58 | cqn1 | sqlinjection.js:31:38:31:41 | cqn1 |
| sqlinjection.js:30:16:30:58 | cds.par ... + book) | sqlinjection.js:30:9:30:58 | cqn1 |
| sqlinjection.js:30:31:30:57 | `SELECT ... `+ book | sqlinjection.js:30:16:30:58 | cds.par ... + book) |
| sqlinjection.js:30:54:30:57 | book | sqlinjection.js:30:31:30:57 | `SELECT ... `+ book |
#select
| sqlinjection.js:13:35:13:39 | query | sqlinjection.js:7:33:7:35 | req | sqlinjection.js:13:35:13:39 | query | This query depends on a $@. | sqlinjection.js:7:33:7:35 | req | user-provided value |
| sqlinjection.js:15:25:15:65 | SELECT. ... book}`) | sqlinjection.js:7:33:7:35 | req | sqlinjection.js:15:25:15:65 | SELECT. ... book}`) | This query depends on a $@. | sqlinjection.js:7:33:7:35 | req | user-provided value |
| sqlinjection.js:18:36:18:41 | query2 | sqlinjection.js:7:33:7:35 | req | sqlinjection.js:18:36:18:41 | query2 | This query depends on a $@. | sqlinjection.js:7:33:7:35 | req | user-provided value |
| sqlinjection.js:20:25:20:63 | SELECT. ... '+book) | sqlinjection.js:7:33:7:35 | req | sqlinjection.js:20:25:20:63 | SELECT. ... '+book) | This query depends on a $@. | sqlinjection.js:7:33:7:35 | req | user-provided value |
| sqlinjection.js:28:38:28:40 | cqn | sqlinjection.js:7:33:7:35 | req | sqlinjection.js:28:38:28:40 | cqn | This query depends on a $@. | sqlinjection.js:7:33:7:35 | req | user-provided value |
| sqlinjection.js:31:38:31:41 | cqn1 | sqlinjection.js:7:33:7:35 | req | sqlinjection.js:31:38:31:41 | cqn1 | This query depends on a $@. | sqlinjection.js:7:33:7:35 | req | user-provided value |
